File: generations/ai-coding-dashboard/lib/a2ui-validator.ts
/**
 * A2UI Protocol Validator (Client-side TypeScript)
 *
 * Validates A2UI messages against v0.8 specification before rendering.
 * Ensures type safety and security constraints.
 */

import {
  A2UIMessage,
  A2UIComponent,
  A2UIMessageType,
  A2UIComponentType,
  A2UI_COMPONENT_TYPES,
  A2UIValidationResult,
  A2UIComponentValidationResult,
  A2UIValidationOptions,
  isA2UIMessageType,
  isA2UIComponentType,
} from '@/types/a2ui';

/**
 * Validate an A2UI message against v0.8 specification
 *
 * @param message - The A2UI message to validate
 * @param options - Validation options
 * @returns Validation result with errors/warnings
 */
export function validateA2UIMessage(
  message: any,
  options: A2UIValidationOptions = {}
): A2UIValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  // Check if message is an object (not an array or null)
  if (typeof message !== 'object' || message === null || Array.isArray(message)) {
    return {
      valid: false,
      errors: ['Message must be an object'],
    };
  }

  // Validate messageType
  if (!message.messageType) {
    errors.push('Missing required field: messageType');
  } else if (!isA2UIMessageType(message.messageType)) {
    errors.push(
      `Invalid messageType: "${message.messageType}". Must be one of: ${Object.values(A2UIMessageType).join(', ')}`
    );
  }

  // Validate components array
  if (!message.components) {
    errors.push('Missing required field: components');
  } else if (!Array.isArray(message.components)) {
    errors.push('Field "components" must be an array');
  } else {
    // Validate each component
    const componentIds = new Set<string>();

    message.components.forEach((component: any, index: number) => {
      const componentResult = validateA2UIComponent(component, index, options);

      if (!componentResult.valid) {
        errors.push(...componentResult.errors);
      }

      // Check for duplicate IDs
      if (component.id) {
        if (componentIds.has(component.id)) {
          errors.push(`Duplicate component ID: "${component.id}"`);
        }
        componentIds.add(component.id);
      }
    });

    // Validate child references
    message.components.forEach((component: any) => {
      if (component.children && Array.isArray(component.children)) {
        component.children.forEach((childId: string) => {
          if (!componentIds.has(childId)) {
            errors.push(
              `Component "${component.id}" references non-existent child: "${childId}"`
            );
          }
        });
      }
    });
  }

  // Check for circular references (children cannot reference ancestors)
  if (message.components && Array.isArray(message.components)) {
    const circularRefErrors = detectCircularReferences(message.components);
    errors.push(...circularRefErrors);
  }

  // Warnings
  if (message.components && message.components.length === 0) {
    warnings.push('Message contains no components');
  }

  const valid = errors.length === 0 && (!options.strict || warnings.length === 0);

  return {
    valid,
    errors,
    warnings: warnings.length > 0 ? warnings : undefined,
    message: valid ? (message as A2UIMessage) : undefined,
  };
}

/**
 * Validate a single A2UI component
 *
 * @param component - The component to validate
 * @param index - Component index (for error messages)
 * @param options - Validation options
 * @returns Component validation result
 */
export function validateA2UIComponent(
  component: any,
  index: number,
  options: A2UIValidationOptions = {}
): A2UIComponentValidationResult {
  const errors: string[] = [];
  const prefix = `Component[${index}]`;

  // Check if component is an object
  if (typeof component !== 'object' || component === null) {
    return {
      valid: false,
      errors: [`${prefix}: Must be an object`],
    };
  }

  // Validate required fields
  if (!component.type) {
    errors.push(`${prefix}: Missing required field "type"`);
  } else if (typeof component.type !== 'string') {
    errors.push(`${prefix}: Field "type" must be a string`);
  } else {
    // Validate component type is in catalog
    const catalog = options.customCatalog || A2UI_COMPONENT_TYPES;

    if (!options.allowUnknownTypes && !catalog.includes(component.type)) {
      errors.push(
        `${prefix}: Invalid component type "${component.type}". Not in registered catalog. Allowed types: ${catalog.join(', ')}`
      );
    }
  }

  if (component.id === undefined || component.id === null) {
    errors.push(`${prefix}: Missing required field "id"`);
  } else if (typeof component.id !== 'string') {
    errors.push(`${prefix}: Field "id" must be a string`);
  } else if (component.id.trim() === '') {
    errors.push(`${prefix}: Field "id" cannot be empty`);
  }

  if (component.props === undefined) {
    errors.push(`${prefix}: Missing required field "props"`);
  } else if (typeof component.props !== 'object' || component.props === null) {
    errors.push(`${prefix}: Field "props" must be an object`);
  }

  // Validate children (optional)
  if (component.children !== undefined) {
    if (!Array.isArray(component.children)) {
      errors.push(`${prefix}: Field "children" must be an array of strings`);
    } else {
      component.children.forEach((child: any, childIndex: number) => {
        if (typeof child !== 'string') {
          errors.push(
            `${prefix}: children[${childIndex}] must be a string (component ID)`
          );
        }
      });
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    componentId: component.id,
  };
}

/**
 * Detect circular references in component tree
 *
 * @param components - List of components to check
 * @returns List of error messages for circular references
 */
function detectCircularReferences(components: A2UIComponent[]): string[] {
  const errors: string[] = [];
  const componentMap = new Map<string, A2UIComponent>();

  // Build component map
  components.forEach((c) => componentMap.set(c.id, c));

  // Check each component for circular references
  components.forEach((component) => {
    if (!component.children) return;

    const visited = new Set<string>();
    const stack = [component.id];

    while (stack.length > 0) {
      const currentId = stack.pop()!;

      if (visited.has(currentId)) {
        errors.push(`Circular reference detected involving component "${currentId}"`);
        break;
      }

      visited.add(currentId);
      const current = componentMap.get(currentId);

      if (current?.children) {
        current.children.forEach((childId) => {
          if (childId === component.id) {
            errors.push(
              `Circular reference: Component "${component.id}" references itself through child "${currentId}"`
            );
          } else {
            stack.push(childId);
          }
        });
      }
    }
  });

  return errors;
}

/**
 * Validate A2UI message and throw if invalid
 *
 * @param message - Message to validate
 * @param options - Validation options
 * @throws Error if validation fails
 */
export function assertValidA2UIMessage(
  message: any,
  options: A2UIValidationOptions = {}
): asserts message is A2UIMessage {
  const result = validateA2UIMessage(message, options);

  if (!result.valid) {
    throw new Error(
      `A2UI validation failed:\n${result.errors.join('\n')}`
    );
  }
}

/**
 * Check if a component type is in the registered catalog
 *
 * @param componentType - Component type to check
 * @param customCatalog - Optional custom catalog (defaults to built-in catalog)
 * @returns true if component type is allowed
 */
export function isComponentTypeAllowed(
  componentType: string,
  customCatalog?: string[]
): boolean {
  const catalog = customCatalog || A2UI_COMPONENT_TYPES;
  return catalog.includes(componentType);
}

/**
 * Get list of all registered component types
 *
 * @param customCatalog - Optional custom catalog
 * @returns Array of allowed component type names
 */
export function getAllowedComponentTypes(customCatalog?: string[]): string[] {
  return customCatalog || [...A2UI_COMPONENT_TYPES];
}

/**
 * Sanitize an A2UI message by removing invalid components
 *
 * @param message - Message to sanitize
 * @param options - Validation options
 * @returns Sanitized message with only valid components
 */
export function sanitizeA2UIMessage(
  message: any,
  options: A2UIValidationOptions = {}
): A2UIMessage {
  if (!message || typeof message !== 'object') {
    return {
      messageType: A2UIMessageType.BEGIN_RENDERING,
      components: [],
    };
  }

  const messageType = isA2UIMessageType(message.messageType)
    ? message.messageType
    : A2UIMessageType.BEGIN_RENDERING;

  const validComponents: A2UIComponent[] = [];

  if (Array.isArray(message.components)) {
    message.components.forEach((component: any, index: number) => {
      const result = validateA2UIComponent(component, index, options);
      if (result.valid) {
        validComponents.push(component);
      }
    });
  }

  return {
    messageType,
    components: validComponents,
    timestamp: message.timestamp,
    metadata: message.metadata,
  };
}

File: generations/ai-coding-dashboard/agent/a2ui_validator.py
"""
A2UI Protocol Validator (Python)

Validates A2UI messages against v0.8 specification.
Ensures JSON structure, required fields, component catalog compliance, and security constraints.
"""

import json
import logging
from typing import Any, Dict, List, Optional, Set, Tuple
from enum import Enum
from datetime import datetime
from pathlib import Path


class A2UIMessageType(str, Enum):
    """Message types supported by A2UI protocol v0.8"""

    BEGIN_RENDERING = "beginRendering"
    SURFACE_UPDATE = "surfaceUpdate"
    DATA_MODEL_UPDATE = "dataModelUpdate"


# Component types registered in A2UI catalog (security constraint)
A2UI_COMPONENT_TYPES = [
    "a2ui.Button",
    "a2ui.Card",
    "a2ui.Text",
    "a2ui.Input",
    "a2ui.Container",
    "a2ui.Grid",
    "a2ui.Badge",
    "a2ui.Divider",
]


class ValidationResult:
    """Result of A2UI validation"""

    def __init__(self, valid: bool = True):
        self.valid = valid
        self.errors: List[str] = []
        self.warnings: List[str] = []
        self.message: Optional[Dict[str, Any]] = None

    def add_error(self, error: str) -> None:
        """Add validation error"""
        self.errors.append(error)
        self.valid = False

    def add_warning(self, warning: str) -> None:
        """Add validation warning"""
        self.warnings.append(warning)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        result = {
            "valid": self.valid,
            "errors": self.errors,
        }
        if self.warnings:
            result["warnings"] = self.warnings
        if self.message:
            result["message"] = self.message
        return result


class A2UIValidator:
    """A2UI Protocol Validator"""

    def __init__(
        self,
        log_file: Optional[str] = None,
        strict: bool = False,
        allow_unknown_types: bool = False,
        custom_catalog: Optional[List[str]] = None,
    ):
        """
        Initialize validator

        Args:
            log_file: Path to validation log file (default: agent/a2ui_validation.log)
            strict: Strict mode - treat warnings as errors
            allow_unknown_types: Allow unknown component types (UNSAFE - testing only)
            custom_catalog: Custom component type allowlist
        """
        self.strict = strict
        self.allow_unknown_types = allow_unknown_types
        self.catalog = custom_catalog if custom_catalog else A2UI_COMPONENT_TYPES

        # Setup logging
        if log_file is None:
            log_file = str(Path(__file__).parent / "a2ui_validation.log")

        self.log_file = log_file
        self.logger = self._setup_logger(log_file)

    def _setup_logger(self, log_file: str) -> logging.Logger:
        """Setup validation logger"""
        logger = logging.getLogger("a2ui_validator")
        logger.setLevel(logging.DEBUG)

        # File handler
        fh = logging.FileHandler(log_file, mode="a")
        fh.setLevel(logging.DEBUG)

        # Formatter
        formatter = logging.Formatter(
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
        )
        fh.setFormatter(formatter)

        logger.addHandler(fh)
        return logger

    def validate_message(self, message: Any) -> ValidationResult:
        """
        Validate A2UI message against v0.8 specification

        Args:
            message: Message to validate (dict or JSON string)

        Returns:
            ValidationResult with errors/warnings
        """
        result = ValidationResult()

        # Parse JSON if string
        if isinstance(message, str):
            try:
                message = json.loads(message)
            except json.JSONDecodeError as e:
                result.add_error(f"Invalid JSON: {str(e)}")
                self.logger.error(f"JSON parsing failed: {str(e)}")
                return result

        # Check if message is a dict
        if not isinstance(message, dict):
            result.add_error("Message must be a JSON object")
            self.logger.error("Message is not a dict")
            return result

        # Validate messageType
        if "messageType" not in message:
            result.add_error("Missing required field: messageType")
        else:
            message_type = message["messageType"]
            if message_type not in [t.value for t in A2UIMessageType]:
                result.add_error(
                    f'Invalid messageType: "{message_type}". '
                    f"Must be one of: {', '.join(t.value for t in A2UIMessageType)}"
                )

        # Validate components array
        if "components" not in message:
            result.add_error("Missing required field: components")
        elif not isinstance(message["components"], list):
            result.add_error("Field 'components' must be an array")
        else:
            components = message["components"]
            component_ids: Set[str] = set()

            # Validate each component
            for index, component in enumerate(components):
                self._validate_component(
                    component, index, result, component_ids
                )

            # Validate child references
            for component in components:
                if "children" in component and isinstance(component["children"], list):
                    for child_id in component["children"]:
                        if child_id not in component_ids:
                            result.add_error(
                                f'Component "{component.get("id", "unknown")}" '
                                f'references non-existent child: "{child_id}"'
                            )

            # Check for circular references
            circular_errors = self._detect_circular_references(components)
            for error in circular_errors:
                result.add_error(error)

        # Warnings
        if "components" in message and len(message["components"]) == 0:
            result.add_warning("Message contains no components")

        # Apply strict mode
        if self.strict and result.warnings:
            result.valid = False

        # Store validated message if valid
        if result.valid:
            result.message = message

        # Log result
        self.logger.info(
            f"Validation {'PASSED' if result.valid else 'FAILED'} "
            f"- Errors: {len(result.errors)}, Warnings: {len(result.warnings)}"
        )
        if not result.valid:
            for error in result.errors:
                self.logger.error(f"  - {error}")

        return result

    def _validate_component(
        self,
        component: Any,
        index: int,
        result: ValidationResult,
        component_ids: Set[str],
    ) -> None:
        """Validate a single component"""
        prefix = f"Component[{index}]"

        # Check if component is a dict
        if not isinstance(component, dict):
            result.add_error(f"{prefix}: Must be an object")
            return

        # Validate required field: type
        if "type" not in component:
            result.add_error(f'{prefix}: Missing required field "type"')
        elif not isinstance(component["type"], str):
            result.add_error(f'{prefix}: Field "type" must be a string')
        else:
            component_type = component["type"]

            # Validate component type is in catalog (SECURITY CHECK)
            if not self.allow_unknown_types and component_type not in self.catalog:
                result.add_error(
                    f'{prefix}: Invalid component type "{component_type}". '
                    f"Not in registered catalog. Allowed types: {', '.join(self.catalog)}"
                )
                self.logger.warning(
                    f"SECURITY: Attempt to use unauthorized component type: {component_type}"
                )

        # Validate required field: id
        if "id" not in component:
            result.add_error(f'{prefix}: Missing required field "id"')
        elif not isinstance(component["id"], str):
            result.add_error(f'{prefix}: Field "id" must be a string')
        elif not component["id"].strip():
            result.add_error(f'{prefix}: Field "id" cannot be empty')
        else:
            component_id = component["id"]

            # Check for duplicate IDs
            if component_id in component_ids:
                result.add_error(f'Duplicate component ID: "{component_id}"')
            else:
                component_ids.add(component_id)

        # Validate required field: props
        if "props" not in component:
            result.add_error(f'{prefix}: Missing required field "props"')
        elif not isinstance(component["props"], dict):
            result.add_error(f'{prefix}: Field "props" must be an object')

        # Validate optional field: children
        if "children" in component:
            children = component["children"]
            if not isinstance(children, list):
                result.add_error(f'{prefix}: Field "children" must be an array of strings')
            else:
                for child_index, child in enumerate(children):
                    if not isinstance(child, str):
                        result.add_error(
                            f"{prefix}: children[{child_index}] must be a string (component ID)"
                        )

    def _detect_circular_references(self, components: List[Dict[str, Any]]) -> List[str]:
        """Detect circular references in component tree"""
        errors: List[str] = []
        component_map = {c["id"]: c for c in components if "id" in c}

        for component in components:
            if "children" not in component or not component.get("id"):
                continue

            visited: Set[str] = set()
            stack = [component["id"]]

            while stack:
                current_id = stack.pop()

                if current_id in visited:
                    errors.append(
                        f'Circular reference detected involving component "{current_id}"'
                    )
                    break

                visited.add(current_id)
                current = component_map.get(current_id)

                if current and "children" in current:
                    for child_id in current["children"]:
                        if child_id == component["id"]:
                            errors.append(
                                f'Circular reference: Component "{component["id"]}" '
                                f'references itself through child "{current_id}"'
                            )
                        else:
                            stack.append(child_id)

        return errors

    def validate_file(self, file_path: str) -> ValidationResult:
        """
        Validate A2UI message from JSON file

        Args:
            file_path: Path to JSON file

        Returns:
            ValidationResult
        """
        self.logger.info(f"Validating file: {file_path}")

        try:
            with open(file_path, "r") as f:
                message = json.load(f)
            return self.validate_message(message)
        except FileNotFoundError:
            result = ValidationResult()
            result.add_error(f"File not found: {file_path}")
            self.logger.error(f"File not found: {file_path}")
            return result
        except json.JSONDecodeError as e:
            result = ValidationResult()
            result.add_error(f"Invalid JSON in file: {str(e)}")
            self.logger.error(f"JSON parsing failed for {file_path}: {str(e)}")
            return result

    def is_component_type_allowed(self, component_type: str) -> bool:
        """
        Check if component type is in registered catalog

        Args:
            component_type: Component type to check

        Returns:
            True if allowed, False otherwise
        """
        return component_type in self.catalog

    def get_allowed_component_types(self) -> List[str]:
        """Get list of all allowed component types"""
        return list(self.catalog)


def validate_a2ui_message(
    message: Any,
    strict: bool = False,
    log_file: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Convenience function to validate A2UI message

    Args:
        message: Message to validate (dict or JSON string)
        strict: Strict mode (treat warnings as errors)
        log_file: Path to log file

    Returns:
        Validation result as dictionary
    """
    validator = A2UIValidator(log_file=log_file, strict=strict)
    result = validator.validate_message(message)
    return result.to_dict()


def validate_a2ui_file(
    file_path: str,
    strict: bool = False,
    log_file: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Convenience function to validate A2UI message from file

    Args:
        file_path: Path to JSON file
        strict: Strict mode (treat warnings as errors)
        log_file: Path to log file

    Returns:
        Validation result as dictionary
    """
    validator = A2UIValidator(log_file=log_file, strict=strict)
    result = validator.validate_file(file_path)
    return result.to_dict()
